### PR TITLE
Add OpenConsole support for Windows

### PR DESCRIPTION
**Pull request by fredizzimo**
If the conpty.dll and OpenConsole.exe files from the Windows Terminal project are placed in the same directory as the alacritty executable they will be used instead of implementation provided by Windows.

This way you don't need to wait for new Windows releases to get bug fixes and new features.

Implements https://github.com/alacritty/alacritty/issues/3889, and using a more recent version of OpenConsole, fixes issues like bad mouse support https://github.com/alacritty/alacritty/issues/1663, curror issues, and probably much more.

In the future Alacritty could probably bundle the files with the release to make sure that a compatible version is used, but right now it's up to the user to compile the Windows Terminal project himself and copy the files.

### CODE REVIEW

**Review by chrisduerr — CHANGES_REQUESTED** [review]
Why would we support experimental software that is not stable yet? If it is meant to be used, it should be provided as part of the Windows operating system.

**Comment by fredizzimo** [comment]
Well, for me it's far more stable, the Windows built in version is basically unusable for me. The mouse does not work, and the cursor is completely messed up in the Neovim terminal. So if I didn't have this choice, I would not use Alacritty at all.

Also openconsole.exe is stable enough to be used by the windows terminal, so I don't see why it wouldn't be stable enough for Alacritty.

**Comment by DHowett** [comment]
So, there's two sides here.

_OpenConsole + conpty.dll_ distributed on their own is experimental only because we have not produced an official package for it.

That hasn't stopped WezTerm (obviously, third-party) and Visual Studio (obviously not third-party!) from doing this. :smile:

We don't officially support it, but we're tracking https://github.com/microsoft/terminal/issues/1130 and https://github.com/microsoft/terminal/issues/3577 as the solution to a bunch of apps distributing them. I'm not against apps taking an unsupported dependency on it, of course. It's just my escape hatch for when things inevitably don't work "quite right" to be able to say it's not supported in an official capacity.

For a bit more background, though: OpenConsole _is_ conhost, just a lot newer because it's not tied to the Windows release cycle. There's one big caveat with redistributing it: if you support WOW64, conpty.dll has to match the application's architecture (obviously, as it's a loadable library) but _openconsole.exe must match the system architecture_. The contract it shares with ConDrv, the kernel driver, is unfortunately bitness-aware.

**Comment by chrisduerr** [comment]
The built-in Windows version being unusable is a Windows problem, not an Alacritty problem. And "stable enough" is hardly a stability guarantee, that's a heuristic at best.

Fact is, the PTY system officially supported by Windows is the one that is shipped with Windows.

I have no problem with people compiling their Alacritty from source to build against an unstable ConPTY interface. That's perfectly fine and everyone is free to do so. But patching Alacritty's stable release to try and support an instable ConPTY version is just incredibly backwards. That's definitely not something I'm interested in fixing inside of Alacritty.

**Comment by ndavidpsi** [comment]
@fredizzimo Since it will not be implemented and I also have the same problem could you provide some steps on how to do the following please?
 
> but right now it's up to the user to compile the Windows Terminal project himself and copy the files.

**Comment by ndavidpsi** [comment]
Do I have to build Windows Terminal from source?

**Comment by ndavidpsi** [comment]
> But patching Alacritty's stable release to try and support an instable ConPTY version is just incredibly backwards. That's definitely not something I'm interested in fixing inside of Alacritty.

Couldn't it be implemented only on Windows builds?

**Comment by fredizzimo** [comment]
Well, this pull request already changes only Windows specific code, the file modified is not compiled on other platforms at all. Furthermore it doesn't modify the behaviour at all, unless you manually copy conpty.dll, and openconsole.exe alongside Alacritty, otherwise it just loads the functionality from kernel32.dll like before, so it's already as nonintrusive as it can get. 

In order to use this pull request, you need to first check it out locally https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally, and if you want then locally merge the latest Alacritty changes into the branch you created.

Then you need to build Alacritty, following the instructions here https://github.com/alacritty/alacritty/blob/master/INSTALL.md, and use the generated exe instead of the normal release.

After that you need to build the Windows terminal, following these instructions https://github.com/microsoft/terminal#developer-guidance. After that, if you copy `conpty.dll` and `openconsole.exe` from the built binaries into the same directory as your custom version of Alacritty that you just built, they should be picked up automatically and the mouse and other issues should be fixed. 

I hope this helps, and you can use Alacritty with working mouse support!

**Comment by ndavidpsi** [comment]
How do I specify the build type for Windows Terminal? I'm building with the powershell method. It did a Debug build, I want a release

**Comment by fredizzimo** [comment]
I built it using Visual Studio myself, but it looks like any additional parameters from the commandline is passed directly to msbuild. So I think `Invoke-OpenConsoleBuild /p:configuration=release /p:platform=x64` should work.

**Comment by ndavidpsi** [comment]
It worked even with the debug version XD. Greatt, thanks it's working fine now

**Comment by DHowett** [comment]
FYI, the debug version of OpenConsole will be very slow. If you continue using it, please make sure you do not file performance issues on either alacritty or microsoft/terminal. 😄

**Comment by ndavidpsi** [comment]
Oh good to have that in mind 😆! Fortunately I can simply replace those two files in the alacritty.exe directory 👍 


(Windows is a pain)

**Comment by ndavidpsi** [comment]
BTW @fredizzimo just to clear up what I did before building alacritty was cloning it from the official repo and then in your pull request I noticed that the only thing you really changed was the `alacritty/alacritty_terminal/src/tty/windows/conpty.rs` file. So what I did was replace the original conpty.rs with yours. Will this cause some issues in the future? Because if I continue to do this for future versions I'll essentially end up with an outdated conpty.rs. Are there some particular lines that I should just change or should I just replace the whole file in the future?

**Comment by fredizzimo** [comment]
That could indeed cause problems in the future, but there's a few ways to handle this.

The safest and easiest way is probably to use Git to track the changes. Create your own branch in your local repository with the changes you made, and when you want to update, you can do (assuming that origin tracks this repository):

1. `git fetch origin` 
2. `git merge origin/master` and resolve any possible conflicts. You can also replace master with a tag to a stable version of  Alacritty, if you prefer to use that instead.

Or alternatively, the manual way, revert these changes before each update, and then manually re-apply them. Assuming that you don't commit these changes to your repository and stay on the master branch
1.  `git checkout alacritty_terminal/src/tty/windows/conpty.rs` (this reverts the file)
2. `git fetch origin`
3. `git merge --ff-only origin/master` (This should never have any conflicts, and the --ff-only flag ensures that you keep following the Alacritty repository exactly)
4. Manually replicate the changes of this pull request into `conpty.rs` 

Note that you can also use `git pull origin master` instead of the fetch/merge combination, if that`s what you are used to.

**Comment by ndavidpsi** [comment]
Sorry, I'm kind of new to git/github. Could you please explain the first option better?

**Comment by ndavidpsi** [comment]
should my first step be `git clone https://github.com/fredizzimo/alacritty`?

**Comment by apazzolini** [comment]
@fredizzimo Thanks for your work here - I followed your instructions and was able to get mouse support working 👍

**Comment by sp1n0r** [comment]
@fredizzimo `conpty.rs` seems to be a lot different in the latest version of alacritty. What should be the procedure now?

**Comment by fredizzimo** [comment]
Yes, it seems to have changed a lot in the commit that removes the WinPTY support. It uses the winapi directly now, instead of dynamically loading the dll, and then getting the  required functions from that.

I can update this pull request with new version, rebased against the latest master, which still removes the WinPty support, but loads the dll dynamically so that it can be overridden by OpenConsole. I'm not sure when I will have time for that, but maybe tomorrow.

BTW, despite this pull request being closed, I'm not holding my breath waiting for Microsoft to fix this. AFAIK Alacritty is the only sotware affected by the bugs. The Windows built in console does not use these features, and their new terminal uses OpenConsole. So for Windows itself these bugs are probably of a very low priority, since they don't break any built-in functionality. Also like mentioned earlier, other similar software like Wezterm and Visual Studio Code have already moved on to OpenConsole due to this.

Furthermore, I'm not even sure if the bugs have been formally reported to the Windows team. It has been reported to Openconsole team, but I don't know if they have reported them further or not. So if the Alacritty team is waiting for this to be fixed by itself, they should report the bugs to Microsoft.

**Comment by DHowett** [comment]
@fredizzimo be careful when you're trying to speak on our behalf :wink:

Those of us working in [the terminal repository](https://github.com/microsoft/terminal) _are_ the Windows console team. Every change made to OpenConsole eventually flows back into Windows. Every change, every fix, every new bug, every enhancement.

conhost is built out of the same code (literally, it's mirrored back into the Windows repository) as OpenConsole.

**Comment by sp1n0r** [comment]
Thanks @fredizzimo :D I'll be waiting.
@DHowett do you know if it is going to be fixed soon on Windows?

**Comment by fredizzimo** [comment]
@DHowett, I'm very sorry about that, I didn't mean to speak on your behalf. And I know that you are the Windows Console team 😄 . 

But it still sounds like there's another team controlling what goes into Windows and when. And that's what I meant, but please correct me if I'm  wrong. The reason why I'm suspecting that is that if everything indeed goes into Windows, then there seems to be either a huge delay, or cherry picking of fixes, since the fixes you did in March did not make it into the 20H2 build. And I was guessing the later, since conhost.exe was updated in that, but at least the mouse support was missing.

Although now that I'm checking the build number, you mentioned build numbers greater than 19500, while 20H2 is only 19042, and the build number only increased by one since the spring H1 update for some reason. So we probably have to wait until the next major release then. 

Or alternatively switch to the Insider builds, which have greater build numbers.

**Comment by DHowett** [comment]
Yeah, that's the unfortunate thing. For the most part, every change made in a Windows release cycle gets released with the next "major" update. We control that, and it's great. Occasionally, like in the case of 19H2 and 20H2, features are released with _targeted backports_. We're not often considered for those (which is a whole other political battle I won't get into :smile:).

My team's working on getting the in-box console host able to delegate work to an externally-installed one (like OpenConsole) so that we're both decoupled from the Windows release cycle and able to deliver ConPTY + traditional console host updates out-of-band.

**Comment by fredizzimo** [comment]
Great, thanks for the clarification.

**Comment by sp1n0r** [comment]
Btw @fredizzimo have you also noticed that if after you edit a fine in vim/nvim you close the editor and the alacritty background becomes grey. It only goes back to normal if you clear the screen.

**Comment by sp1n0r** [comment]
Any progress?

**Comment by fredizzimo** [comment]
Sorry for the delay. But I have rebased the pull request against the latest Alacritty master now and updated the file in my branch https://github.com/fredizzimo/alacritty/tree/add_openconsole_support.

Unfortunately this pull request was not been automatically updated, since it's closed, so to use my new version, you can either do it the lazy way, and copy the file directly from my github fork https://github.com/fredizzimo/alacritty/blob/add_openconsole_support/alacritty_terminal/src/tty/windows/conpty.rs. Alternatively you can add a remote and check-out the branch with git
`git remote add fredizzimo https://github.com/fredizzimo/alacritty`
`git fetch fredizzimo`
`git checkout fredizzimo/add_openconsole_support`

@sp1n0r , yes I also have the background problems, but I have never bothered to check what's causing it or if it's fixable. Actually after trying it I'm not sure if I have the problem now or not, since the background seemed to be normal and I can't remember when I actually saw it the last time. It might also be that some of the nightly Neovim versions have fixed it. But I'm also using the same background in the terminal as Vim, so that also helps.

**Comment by sp1n0r** [comment]
Thanks!! We'll regarding the background problem it seems to happen every time I edit something, I'm using one of the latest builds of neovim as well (1 week old I think) and I have the background color set to NONE in Neovim, so for me both the Terminal and the editor have the same black background, alas the grey background issue occurs the same. Let's hope this newer version of alacritty solves it somehow. Again, thanks for updating the file 😄

**Comment by fredizzimo** [comment]
I have rebased my branch against the v0.8.0 version of alacritty.

Unfortunately Windows 21H1 still does not include the OpenConsole fixes and with the upcoming release of Windows 11, I doubt they will ever be included, unortunately. 

@DHowett, is there anything you can do to make the fixes included in the 21H2 version of Windows? 

Alternatively @chrisduerr, could you please re-consider and re-open the pull request? In addition to this pull request, I also think that the Alacritty release should include a release version of openconsole that is confirmed to be working, just like many other terminal emulators do on Windows. That way it should just work without any manual steps. 

I have personally been running Alacritty with OpenConsole for over a year, without any issues, and I have been using this code since I made the pull request, also without any issues.

On the other hand Alacritty is **unusable** on Windows without using OpenConsole. It might work for very basic terminal usage, but Neovim for example is completely broken.

**Comment by chrisduerr** [comment]
> Alternatively @chrisduerr, could you please re-consider and re-open the pull request?

No.

**Comment by andrewvaughanj** [comment]
> On the other hand Alacritty is unusable on Windows without using OpenConsole. It might work for very basic terminal usage, but Neovim for example is completely broken.
>

As someone who spends upwards of 8 hours a day, 5 days a week working with Alacritty on Windows, I can tell you that it is very much *not* "completely broken" (as long as you've patched around copying-pasting, which admittedly is broken due to ConPTY).

**Comment by fredizzimo** [comment]
That's strange, because I use also use Alacritty every day at work too, and I have not run into the double line paste issue. But maybe that's because Neovim handles that correctly, at least with the patched OpenConsole.exe. But I did see the issue when I tried it in the terminal just a while ago.

However, I can't live with the completely broken mouse support https://github.com/alacritty/alacritty/issues/1663. (yes, I need to be able to scroll with my mouse wheel, when I browse the code, even if I normally use the keyboard only to write it)

or cursor issues like this https://github.com/alacritty/alacritty/issues/4573

**Comment by mks1508** [comment]
@fredizzimo After try this patch for a week... I go back here and wanna say ... Thank you so much.
Patched alacritty is so smooth and stable ... the performace is even better using patched newest alacritty instead v0.8.0.
Some random input lags/delays have been completely gone! 👍

**Comment by Viqsi** [comment]
> > Alternatively @chrisduerr, could you please re-consider and re-open the pull request?
> 
> No.

I'm not sure I get the point of this decision, then - particularly given WinPTY support's removal a year ago. If you're not going to support the old version of Windows console management in WinPTY, and you're not going to support current Windows console management in ConPTY (like literally everybody else in this space does, up to and including Microsoft themselves), then Alacritty on Windows is arguably hopelessly crippled, and one has to wonder why the Windows port exists at all.

**Comment by andrewvaughanj** [comment]
I'm a proponent of this change, but:

> Windows is arguably hopelessly crippled

I think that's likely to be somewhat unfair (as someone who uses Alacritty for > 50 hours a week on Windows)

**Comment by Viqsi** [comment]
> I'm a proponent of this change, but:
> 
> > Windows is arguably hopelessly crippled
> 
> I think that's likely to be somewhat unfair (as someone who uses Alacritty for > 50 hours a week on Windows)

Fair point; I guess I'm just grumpier than usual at the moment having to reimplement a workaround for the fourth time in two weeks and thus magnifying the issue in my head. :)

It still strikes me as unnecessarily adversarial to just outright refuse the current state of Windows terminal emulation, tho. Especially if the patches are being provided and don't affect other platforms at all...

**Comment by chrisduerr** [comment]
> you're not going to support current Windows console management in ConPTY

We support the current Windows console management in ConPTY, using Microsoft's official API.

**Comment by Viqsi** [comment]
> > you're not going to support current Windows console management in ConPTY
> 
> We support the current Windows console management in ConPTY, using Microsoft's official API.

Microsoft themselves don't use that version of the API, though. They use current OpenConsole instead in their terminal. As does virtually everyone else producing terminal emulators for Windows; those who don't use it are using the Windows console API instead (support for which was recently dropped in Alacritty). So far as I have been able to tell by surveying Windows terminal emulators, Alacritty is the sole and exclusive holdout.

**Comment by chrisduerr** [comment]
> Microsoft themselves don't use that version of the API, though. They use current OpenConsole instead in their terminal. As does virtually everyone else producing terminal emulators for Windows; those who don't use it are using the old Windows console API instead (support for which was, as I understand it, recently dropped in Alacritty). So far as I have been able to tell by surveying Windows terminal emulators, Alacritty is the sole and exclusive holdout.

And the issue here is that Alacritty uses the **standard** public API, not that Microsoft is incapable of shipping a usable ConPTY API? I'm not going to lose any sleep over issues that Microsoft can easily solve themselves whenever they feel like it.

**Comment by andrewvaughanj** [comment]
@Viqsi if you care, just build Alacritty from source with @fredizzimo's patches applied -- that's what I do.

@chrisduerr isn't losing any sleep on missing this functionality, and I don't lose any sleep on building my own Alacritty from source to get these patches ¯\\\_(ツ)\_/¯

**Comment by awoodx** [comment]
thanks @fredizzimo! applied your patch and now the mouse wheel works. also a bunch of other issues seemed fixed too.. had an issue where the cursor would flicker around the screen when using FZF. that seems fixed by the patch.

**Comment by Grueslayer** [comment]
Just need to change the patch a little bit to use windows_sys instead of winapi and works fine. Don't understand the stubborness of @chrisduerr for rejecting the patch: It does no harm at all. As long you have no conpty.dll in path (which you will not have, cause in Windows Terminal it is statically linked) it behaves like before. You can use Alacritty only with commandline programs where there is no terminal interaction except normal output.

So we'll need to build our own terminal or go with any other using the old winpty (ConEmu still runs fine) or using the Windows Terminal ConPTY (like WizTerm does).

**Comment by aytey** [comment]
@Grueslayer can you share your updated patch?

**Comment by aytey** [comment]
> Don't understand the stubborness of @chrisduerr for rejecting the patch: It does no harm at all. 
>

@chrisduerr maintains Alacritty; he's welcome to accept and reject patches he feels don't match the ethos and direction of Alacritty. Chris simply feels that we're "working around something we shouldn't have to work around", and doesn't want to "litter" the Alacritty tree with such workarounds.

Honestly, I do not disagree with him, so I consider it "fine" to maintain my own set of patches against Alacritty to help me use it better on Windows.

**Comment by Grueslayer** [comment]
I was searching for a terminal replacing `iTerm2` on Mac, `ConEmu` on Windows and `Kitty` on BSD/Linux. 
Alacritty claims to be, but it is really in no shape for using it on Windows (beside the multiple bugs here which could be fixed with this patch, I discovered many more in Windows in my one day usage).

Maintainer can do what he/she wants, I don't understand whats about maintainig a crippled windows version instead of trying to fix it (or drop it). But I don't need to, so I'll go with any other stable available terminal.

@aytey Here is your requested patch:

```diff
diff --git a/alacritty_terminal/src/tty/windows/conpty.rs b/alacritty_terminal/src/tty/windows/conpty.rs
index 3f6349de..2b9439d2 100644
--- a/alacritty_terminal/src/tty/windows/conpty.rs
+++ b/alacritty_terminal/src/tty/windows/conpty.rs
@@ -4,11 +4,15 @@ use std::{mem, ptr};
 
 use mio_anonymous_pipes::{EventedAnonRead, EventedAnonWrite};
 
-use windows_sys::core::PWSTR;
+use windows_sys::core::{PWSTR, HRESULT};
 use windows_sys::Win32::Foundation::{HANDLE, S_OK};
 use windows_sys::Win32::System::Console::{
-    ClosePseudoConsole, CreatePseudoConsole, ResizePseudoConsole, COORD, HPCON,
+    COORD, HPCON
 };
+use windows_sys::Win32::System::LibraryLoader::{
+    GetModuleHandleA, GetProcAddress, LoadLibraryA
+};
+
 use windows_sys::Win32::System::Threading::{
     CreateProcessW, InitializeProcThreadAttributeList, UpdateProcThreadAttribute,
     EXTENDED_STARTUPINFO_PRESENT, PROCESS_INFORMATION, PROC_THREAD_ATTRIBUTE_PSEUDOCONSOLE,
@@ -20,9 +24,55 @@ use crate::event::{OnResize, WindowSize};
 use crate::tty::windows::child::ChildExitWatcher;
 use crate::tty::windows::{cmdline, win32_string, Pty};
 
+/// Dynamically load the pseudoconsole API from either kernel32.dll
+/// or conpty.dll, depending on if the latter exists or not. The conpty.dll
+/// from the Windows Terminal project supports loading OpenConsole.exe, which
+/// has has many improvements and bugfixes compared to the standard conpty that
+/// ships with Windows.
+///
+/// Just copy the conpty.dll and OpenConsole.exe files to the same
+/// directory as the Alacritty.exe to use it.
+///
+/// The field names are deliberately PascalCase as this matches
+/// the defined symbols in kernel32 and also is the convention
+/// that the `winapi` crate follows.
+#[allow(non_snake_case)]
+struct ConptyApi {
+    CreatePseudoConsole:
+        unsafe extern "system" fn(COORD, HANDLE, HANDLE, u32, *mut HPCON) -> HRESULT,
+    ResizePseudoConsole: unsafe extern "system" fn(HPCON, COORD) -> HRESULT,
+    ClosePseudoConsole: unsafe extern "system" fn(HPCON),
+}
+
+impl ConptyApi {
+    /// Load the API or None if it cannot be found.
+    pub fn new() -> Option<Self> {
+        // Unsafe because windows API calls.
+        unsafe {
+            let mut hmodule = LoadLibraryA("conpty.dll\0".as_ptr() as _);
+            if hmodule == 0x0 {
+                // Otherwise just use the standard conpty from kernel32.
+                hmodule = GetModuleHandleA("kernel32\0".as_ptr() as _);
+            }
+            assert!(hmodule != 0x0);
+
+            let cpc = GetProcAddress(hmodule, "CreatePseudoConsole\0".as_ptr() as _);
+            let rpc = GetProcAddress(hmodule, "ResizePseudoConsole\0".as_ptr() as _);
+            let clpc = GetProcAddress(hmodule, "ClosePseudoConsole\0".as_ptr() as _);
+
+            Some(Self {
+                CreatePseudoConsole: mem::transmute(cpc),
+                ResizePseudoConsole: mem::transmute(rpc),
+                ClosePseudoConsole: mem::transmute(clpc),
+            })
+        }
+    }
+}
+
 /// RAII Pseudoconsole.
 pub struct Conpty {
     pub handle: HPCON,
+    api: ConptyApi,
 }
 
 impl Drop for Conpty {
@@ -31,7 +81,7 @@ impl Drop for Conpty {
         // conout pipe has already been dropped by this point.
         //
         // See PR #3084 and https://docs.microsoft.com/en-us/windows/console/closepseudoconsole.
-        unsafe { ClosePseudoConsole(self.handle) }
+        unsafe { (self.api.ClosePseudoConsole)(self.handle) }
     }
 }
 
@@ -39,6 +89,7 @@ impl Drop for Conpty {
 unsafe impl Send for Conpty {}
 
 pub fn new(config: &PtyConfig, window_size: WindowSize) -> Option<Pty> {
+    let api = ConptyApi::new()?;
     let mut pty_handle: HPCON = 0;
 
     // Passing 0 as the size parameter allows the "system default" buffer
@@ -50,7 +101,7 @@ pub fn new(config: &PtyConfig, window_size: WindowSize) -> Option<Pty> {
 
     // Create the Pseudo Console, using the pipes.
     let result = unsafe {
-        CreatePseudoConsole(
+        (api.CreatePseudoConsole)(
             window_size.into(),
             conin_pty_handle.into_raw_handle() as HANDLE,
             conout_pty_handle.into_raw_handle() as HANDLE,
@@ -158,7 +209,7 @@ pub fn new(config: &PtyConfig, window_size: WindowSize) -> Option<Pty> {
     let conout = EventedAnonRead::new(conout);
 
     let child_watcher = ChildExitWatcher::new(proc_info.hProcess).unwrap();
-    let conpty = Conpty { handle: pty_handle as HPCON };
+    let conpty = Conpty { handle: pty_handle as HPCON, api };
 
     Some(Pty::new(conpty, conout, conin, child_watcher))
 }
@@ -170,7 +221,7 @@ fn panic_shell_spawn() {
 
 impl OnResize for Conpty {
     fn on_resize(&mut self, window_size: WindowSize) {
-        let result = unsafe { ResizePseudoConsole(self.handle, window_size.into()) };
+        let result = unsafe { (self.api.ResizePseudoConsole)(self.handle, window_size.into()) };
         assert_eq!(result, S_OK);
     }
 }
```

**Comment by DerpyDashie** [comment]
I want to note for anyone who wants to apply this, if you don't want to build the windows terminal, you can copy OpenConsole.exe and conpty.dll from [wezterm](https://github.com/wez/wezterm/tree/main/assets/windows/conhost), they also include them in their releases

**Comment by proski** [comment]
Thank you, @Grueslayer! Your patch applies to the master branch, but I also had to add the `Win32_System_LibraryLoader` feature to `alacritty_terminal/Cargo.toml`. It's working perfectly for me! I've made a branch in my fork: https://github.com/proski/alacritty/tree/local-conpty

**Comment by proski** [comment]
I've cleaned up the patch significantly. All Windows API calls are checked. `transmute` is done on pointers, not on `Option`. There are no casts. The `s!` macro is used for strings. The fallback is the original API, not symbols loaded from `kernel32.dll` - that eliminates any possibility for the fallback to fail. The comments have been updated. It should be good enough for submission, but it looks like the objections are not based on the code quality, so I won't submit it.

It's been a good learning opportunity for me and I have a functional terminal emulator now.

**Comment by kchibisov** [comment]
@proski could you send your patch upstream?

**Pull request closed without merging**
